modules/create_timepoint/php/ajaxtimepoint.class.inc

```diff
@@ -0,0 +1,285 @@
+            add new subprojects and associate them with projects.';
+        }
+        $values['subproject'] = array_intersect_key(
+            \Utility::getSubprojectList(),
+            $candidate->getValidSubprojects()
+        );
+
+        // Retrieve visit labels.
+        $visit_options      = array();
+        $visitLabelSettings = $config->getSetting('visitLabel');
+        foreach (
+            \Utility::toArray($visitLabelSettings) as $visitLabel
+        ) {
+            if (!empty($values['subproject'])) {
+                $visitLabel       = \Utility::toArray($visitLabel)[0];
+                $labelOptions[''] = null;
+                $items            = \Utility::toArray(
+                    $visitLabel['labelSet']['item']
+                );
+                foreach ($items as $item) {
+                    $item = \Utility::toArray($item)[0];
+                    $labelOptions[$item['@']['value']] = $item['#'];
+                }
+                $visit_options[$visitLabel['@']['subprojectID']]
+                    = array_filter($labelOptions);
+            }
+        }
+        $values['visit'] = $visit_options;
+        $values['psc']   = array_intersect(
+            \Utility::getSiteList(),
+            \User::singleton()->getSiteNames()
+        );
+        if (!empty($errors)) {
+            $values['errors'] = $errors;
+        }
+        if (empty($values['psc'])) {
+            unset($values['psc']);
+        }
+
+        return new \LORIS\Http\Response\JsonResponse(
+            $values
+        );
+    }
+
+    /**
+     * Processes the values & saves to database and return a json response.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handlePOST(ServerRequestInterface $request) : ResponseInterface
+    {
+        // Parse POST request body.
+        $values = $request->getParsedBody();
+
+        // User permission verification.
+        $values = $this->verifyPermissions($request, $values);
+        if (isset($values['error'])) {
+            switch ($values['error']) {
+            case 'BadRequest':
+                return new \LORIS\Http\Response\BadRequest(
+                    'Failed to retrieve data for create timepoint.'
+                );
+            case 'Forbidden':
+                return new \LORIS\Http\Response\Forbidden(
+                    'Forbidden'
+                );
+            default:
+                break;
+            }
+        }
+
+        if (isset($values['command']) && $values['command'] == 'create') {
+            $errors = $this->validate($values);
+            // convert site entered to a \Site object
+            $site = null;
+            $user = \User::singleton();
+            $user_list_of_sites = $user->getCenterIDs();
+            $num_sites          = count($user_list_of_sites);
+            if ($num_sites == 1) {
```

Comment:
Could you make the React app always send the Site even when the user as only one? Maybe just don't display the fields when it's the case.
We would then be able to avoid this "auto-select" feature in the backend.

Reply:
@maltheism I the POST request body, there should be a site.

Reply:
@xlecours you're right. I'm still confused on what you want me to do because won't I still need to use $site = \Site::singleton($user_list_of_sites[0]); for retrieving the site object that's used for creating the timepoint?

Reply:
`$site = \Site::singleton(intval($values['psc']));` would be enough. Because `$values` will always have a `psc` otherwise a BadRequest would have been sent earlier.

the only scenario where this would fail is if the user is only part of a study site. The array intersect at line#136. in the handleGET, would not send a psc to the front end because Utility::getSiteList would remove the DCC sites from the list. Which would create a weird behavior with `$site = \Site::singleton($user_list_of_sites[0]);` (I would register the session as part of the study_site)